### ISSUE TITLE
Slow query against SYS.ALL_INDEXES in OracleDatabase

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14014]
Since a few months, there has been an excessively slow query in `OracleDatabase`, in our integration tests:

```
09:01:40,608 [WARNING] Slow SQL                 : Total: 01:14
09:01:40,613 [WARNING] Slow SQL                 : jOOQ Meta executed a slow query (slower than 5 seconds, configured by configuration/generator/database/logSlowQueriesAfterSeconds)
```

The query is this:

```sql
select
  SYS.ALL_INDEXES.OWNER,
  SYS.ALL_INDEXES.INDEX_NAME,
  SYS.ALL_INDEXES.TABLE_OWNER,
  SYS.ALL_INDEXES.TABLE_NAME,
  decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS
from SYS.ALL_INDEXES
where (
  SYS.ALL_INDEXES.OWNER in (
    'TEST', 'MULTI_SCHEMA', 'USR_2522', 'SYS'
  )
  and SYS.ALL_INDEXES.TABLE_OWNER in (
    'TEST', 'MULTI_SCHEMA', 'USR_2522', 'SYS'
  )
  and SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL'
  and (SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((
    select SYS.ALL_CONSTRAINTS.OWNER, SYS.ALL_CONSTRAINTS.TABLE_NAME, SYS.ALL_CONSTRAINTS.CONSTRAINT_NAME
    from SYS.ALL_CONSTRAINTS
  ))
)
order by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME
```

Given the low importance of generating index meta data, there's certainly something that should be done.

Investigation will follow.

**Comment by lukaseder** [comment]
Explain plan looks OK:

```
|PLAN_TABLE_OUTPUT                                                                                                     |
|----------------------------------------------------------------------------------------------------------------------|
|SQL_ID  7xwhwrzv5vpg3, child number 0                                                                                 |
|-------------------------------------                                                                                 |
|select SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME,                                                             |
|SYS.ALL_INDEXES.TABLE_OWNER, SYS.ALL_INDEXES.TABLE_NAME,                                                              |
|decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS from                                                    |
|SYS.ALL_INDEXES where (SYS.ALL_INDEXES.OWNER in (:1 , :2 , :3 , :4 )                                                  |
|and SYS.ALL_INDEXES.TABLE_OWNER in (:5 , :6 , :7 , :8 ) and                                                           |
|SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL' and (SYS.ALL_INDEXES.OWNER,                                                     |
|SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((select                                               |
|SYS.ALL_CONSTRAINTS.OWNER, SYS.ALL_CONSTRAINTS.TABLE_NAME,                                                            |
|SYS.ALL_CONSTRAINTS.CONSTRAINT_NAME from SYS.ALL_CONSTRAINTS))) order                                                 |
|by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME -- SQL rendered                                                  |
|with a free trial version of jOOQ 3.18.0-SNAPSHOT                                                                     |
|                                                                                                                      |
|Plan hash value: 82967922                                                                                             |
|                                                                                                                      |
|----------------------------------------------------------------------------------------------------------------      |
|| Id  | Operation                                | Name                    | E-Rows |  OMem |  1Mem | Used-Mem |      |
|----------------------------------------------------------------------------------------------------------------      |
||   0 | SELECT STATEMENT                         |                         |        |       |       |          |      |
||   1 |  SORT ORDER BY                           |                         |      1 |  9216 |  9216 | 8192  (0)|      |
||*  2 |   FILTER                                 |                         |        |       |       |          |      |
||   3 |    NESTED LOOPS                          |                         |      1 |       |       |          |      |
||   4 |     NESTED LOOPS OUTER                   |                         |      1 |       |       |          |      |
||   5 |      NESTED LOOPS                        |                         |      1 |       |       |          |      |
||   6 |       NESTED LOOPS OUTER                 |                         |      1 |       |       |          |      |
||   7 |        NESTED LOOPS OUTER                |                         |      1 |       |       |          |      |
||   8 |         NESTED LOOPS OUTER               |                         |      1 |       |       |          |      |
||   9 |          NESTED LOOPS                    |                         |      1 |       |       |          |      |
||  10 |           NESTED LOOPS                   |                         |     83 |       |       |          |      |
||  11 |            INLIST ITERATOR               |                         |        |       |       |          |      |
||  12 |             TABLE ACCESS BY INDEX ROWID  | USER$                   |      4 |       |       |          |      |
||* 13 |              INDEX UNIQUE SCAN           | I_USER1                 |      4 |       |       |          |      |
||* 14 |            INDEX RANGE SCAN              | I_OBJ2                  |     21 |       |       |          |      |
||* 15 |           TABLE ACCESS CLUSTER           | IND$                    |      1 |       |       |          |      |
||* 16 |            INDEX UNIQUE SCAN             | I_OBJ#                  |      1 |       |       |          |      |
||  17 |          TABLE ACCESS CLUSTER            | TS$                     |      1 |       |       |          |      |
||* 18 |           INDEX UNIQUE SCAN              | I_TS#                   |      1 |       |       |          |      |
||  19 |         TABLE ACCESS CLUSTER             | SEG$                    |      1 |       |       |          |      |
||* 20 |          INDEX UNIQUE SCAN               | I_FILE#_BLOCK#          |      1 |       |       |          |      |
||* 21 |        INDEX RANGE SCAN                  | I_OBJ1                  |      1 |       |       |          |      |
||* 22 |       TABLE ACCESS BY INDEX ROWID BATCHED| OBJ$                    |      1 |       |       |          |      |
||* 23 |        INDEX RANGE SCAN                  | I_OBJ1                  |      1 |       |       |          |      |
||* 24 |      INDEX RANGE SCAN                    | I_USER2                 |      1 |       |       |          |      |
||* 25 |     TABLE ACCESS CLUSTER                 | USER$                   |      1 |       |       |          |      |
||* 26 |      INDEX UNIQUE SCAN                   | I_USER#                 |      1 |       |       |          |      |
||* 27 |    FILTER                                |                         |        |       |       |          |      |
||  28 |     PARTITION LIST ALL                   |                         |     15 |       |       |          |      |
||* 29 |      EXTENDED DATA LINK FULL             | INT$INT$DBA_CONSTRAINTS |     15 |       |       |          |      |
||  30 |     NESTED LOOPS SEMI                    |                         |      1 |       |       |          |      |
||* 31 |      FIXED TABLE FULL                    | X$KZSRO                 |      2 |       |       |          |      |
||* 32 |      INDEX RANGE SCAN                    | I_OBJAUTH2              |      1 |       |       |          |      |
||* 33 |    TABLE ACCESS CLUSTER                  | TAB$                    |      1 |       |       |          |      |
||* 34 |     INDEX UNIQUE SCAN                    | I_OBJ#                  |      1 |       |       |          |      |
||  35 |    NESTED LOOPS SEMI                     |                         |      1 |       |       |          |      |
||* 36 |     FIXED TABLE FULL                     | X$KZSRO                 |      2 |       |       |          |      |
||* 37 |     INDEX RANGE SCAN                     | I_OBJAUTH2              |      1 |       |       |          |      |
|----------------------------------------------------------------------------------------------------------------      |
|                                                                                                                      |
|Predicate Information (identified by operation id):                                                                   |
|---------------------------------------------------                                                                   |
|                                                                                                                      |
|   2 - filter(( IS NULL AND ("IO"."TYPE#"<>2 OR =1) AND ("IO"."OWNER#"=USERENV('SCHEMAID') OR                         |
|              ORA_CHECK_SYS_PRIVILEGE("IO"."OWNER#","IO"."TYPE#")=1 OR  IS NOT NULL)))                                |
|  13 - access(("IU"."NAME"=:5 OR "IU"."NAME"=:6 OR "IU"."NAME"=:7 OR "IU"."NAME"=:8))                                 |
|  14 - access("IO"."OWNER#"="IU"."USER#" AND "IO"."TYPE#"=2)                                                          |
|       filter("IO"."TYPE#"=2)                                                                                         |
|  15 - filter((INTERNAL_FUNCTION("I"."TYPE#") AND DECODE(BITAND("I"."PROPERTY",16),0,'','FUNCTION-BASED               |
|              ')||DECODE("I"."TYPE#",1,'NORMAL'||DECODE(BITAND("I"."PROPERTY",4),0,'',4,'/REV'),2,'BITMAP',3,'CLUSTER'|
|              ,4,'IOT - TOP',5,'IOT - NESTED',6,'SECONDARY',7,'ANSI',8,'LOB',9,'DOMAIN')='NORMAL' AND                 |
|              BITAND("I"."FLAGS",4096)=0))                                                                            |
|  16 - access("I"."BO#"="IO"."OBJ#")                                                                                  |
|  18 - access("I"."TS#"="TS"."TS#")                                                                                   |
|  20 - access("I"."TS#"="S"."TS#" AND "I"."FILE#"="S"."FILE#" AND "I"."BLOCK#"="S"."BLOCK#")                          |
|  21 - access("I"."INDMETHOD#"="ITO"."OBJ#")                                                                          |
|  22 - filter(BITAND("O"."FLAGS",128)=0)                                                                              |
|  23 - access("O"."OBJ#"="I"."OBJ#")                                                                                  |
|  24 - access("ITO"."OWNER#"="ITU"."USER#")                                                                           |
|  25 - filter(("U"."NAME"=:1 OR "U"."NAME"=:2 OR "U"."NAME"=:3 OR "U"."NAME"=:4))                                     |
|  26 - access("U"."USER#"="O"."OWNER#")                                                                               |
|  27 - filter(("OWNER"=SYS_CONTEXT('USERENV','CURRENT_USER') OR                                                       |
|              ORA_CHECK_SYS_PRIVILEGE("OWNERID","OBJECT_TYPE#")=1 OR  IS NOT NULL))                                   |
|  29 - filter((("INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=4 OR                                                         |
|              ("INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=2 AND "INT$INT$DBA_CONSTRAINTS"."ORIGIN_CON_ID"=TO_NUMBER(SYS_|
|              CONTEXT('USERENV','CON_ID')))) AND LNNVL("OWNER"<>:B1) AND LNNVL("TABLE_NAME"<>:B2) AND                 |
|              LNNVL("CONSTRAINT_NAME"<>:B3)))                                                                         |
|  31 - filter(("CON_ID"=0 OR "CON_ID"=3))                                                                             |
|  32 - access("GRANTEE#"="KZSROROL" AND "OBJ#"=OBJ_ID(:B1,:B2,:B3,:B4))                                               |
|  33 - filter(BITAND("T"."PROPERTY",36893488147419103232)=0)                                                          |
|  34 - access("T"."OBJ#"=:B1)                                                                                         |
|  36 - filter(("CON_ID"=0 OR "CON_ID"=3))                                                                             |
|  37 - access("GRANTEE#"="KZSROROL" AND "OBJ#"=:B1)                                                                   |
|                                                                                                                      |
|Note                                                                                                                  |
|-----                                                                                                                 |
|   - Warning: basic plan statistics not available. These are only collected when:                                     |
|       * hint 'gather_plan_statistics' is used for the statement or                                                   |
|       * parameter 'statistics_level' is set to 'ALL', at session or system level                                     |
|                                                                                                                      |
```

But a query on `v$sql` shows an excessive amount of executions to this auxiliary query, probably from a PL/SQL function:

```sql
select 
  sql_id,
  child_number,
  elapsed_time / 1000000.0 as et, 
  executions as ex,
  cast(elapsed_time / 1000000.0 / nullif(executions, 0) as number(20, 3)) as tpe,
  sql_text
  -- s.*
from v$sql s
where 1 = 1
-- and parsing_schema_name = 'TEST'
and module = 'JDBC Thin Client'
order by elapsed_time desc
;
```

Result:

```
|SQL_ID       |CHILD_NUMBER|ET       |EX   |TPE   |SQL_TEXT                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|-------------|------------|---------|-----|------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|7xwhwrzv5vpg3|0           |74.676678|1    |74.677|select SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME, SYS.ALL_INDEXES.TABLE_OWNER, SYS.ALL_INDEXES.TABLE_NAME, decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS from SYS.ALL_INDEXES where (SYS.ALL_INDEXES.OWNER in (:1 , :2 , :3 , :4 ) and SYS.ALL_INDEXES.TABLE_OWNER in (:5 , :6 , :7 , :8 ) and SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL' and (SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((select SYS.ALL_CONSTRAINTS.OWNER, SYS.ALL_CONSTRAINTS.TABLE_NAME, SYS.ALL_CONSTRAINTS.CONSTRAINT_NAME from SYS.ALL_CONSTRAINTS))) order by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT                                                                                                                                                                                               |
|6kd3rkh98hc5y|0           |63.507871|1,597|0.04  |SELECT /*+ NO_STATEMENT_QUEUING RESULT_CACHE (SYSOBJ=TRUE) OPT_PARAM('_ENABLE_VIEW_PDB', 'FALSE') */ "OWNER","OWNERID","CONSTRAINT_NAME","TABLE_NAME","OBJECT_ID","OBJECT_TYPE#","ORIGIN_CON_ID" FROM NO_CROSS_CONTAINER("SYS"."INT$INT$DBA_CONSTRAINTS") "INT$INT$DBA_CONSTRAINTS" WHERE ("INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=4 OR "INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=2 AND "INT$INT$DBA_CONSTRAINTS"."ORIGIN_CON_ID"=TO_NUMBER('3')) AND 1=1 AND 1=1 AND 1=1                                                                                                                                                                                                                                                                                                                                                                                                                                              |
|23mrzz17v4hhq|0           |25.229699|438  |0.058 |select a.IN_OUT, a.ARGUMENT_NAME, a.DATA_TYPE, a.CHAR_LENGTH, a.DATA_PRECISION, a.DATA_SCALE, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.OWNER else a.TYPE_OWNER end TYPE_OWNER, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.PACKAGE_NAME else a.TYPE_NAME end TYPE_NAME, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then (a.OBJECT_NAME || nvl2(a.ARGUMENT_NAME, ('_' || a.ARGUMENT_NAME), '')) else a.TYPE_SUBNAME end TYPE_SUBNAME, a.POSITION, a.DEFAULTED from SYS.ALL_ARGUMENTS a where (a.OWNER = :1  and a.OBJECT_NAME = :2  and a.PACKAGE_NAME = :3  and decode(a.OVERLOAD, :4 , 1, 0) = 1 and a.DATA_LEVEL = :5  and a.DATA_TYPE is not null) order by a.POSITION asc -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT|
```

Formatted query:

```sql
SELECT /*+ NO_STATEMENT_QUEUING RESULT_CACHE (SYSOBJ=TRUE) OPT_PARAM('_ENABLE_VIEW_PDB', 'FALSE') */ 
  "OWNER","OWNERID","CONSTRAINT_NAME","TABLE_NAME","OBJECT_ID","OBJECT_TYPE#","ORIGIN_CON_ID" 
FROM NO_CROSS_CONTAINER("SYS"."INT$INT$DBA_CONSTRAINTS") "INT$INT$DBA_CONSTRAINTS" 
WHERE ("INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=4 
  OR "INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=2 
  AND "INT$INT$DBA_CONSTRAINTS"."ORIGIN_CON_ID"=TO_NUMBER('3')) 
  AND 1=1 AND 1=1 AND 1=1
```

**Comment by lukaseder** [comment]
Looks like it's originating from here:

```
|* 27 |    FILTER                                |                         |        |       |       |          |      |
||  28 |     PARTITION LIST ALL                   |                         |     15 |       |       |          |      |
||* 29 |      EXTENDED DATA LINK FULL             | INT$INT$DBA_CONSTRAINTS |     15 |       |       |          |      |
```

**Comment by lukaseder** [comment]
It seems there isn't a `HASH JOIN ANTI` operation between the `ALL_INDEXES` and `ALL_CONSTRAINTS` tables, as expected. Instead, some sort of nested loop is run.

Here an execution plan with `ALTER SYSTEM SET statistics_level = all`:

```
|PLAN_TABLE_OUTPUT                                                                                                                                             |
|--------------------------------------------------------------------------------------------------------------------------------------------------------------|
|SQL_ID  7xwhwrzv5vpg3, child number 1                                                                                                                         |
|-------------------------------------                                                                                                                         |
|select SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME,                                                                                                     |
|SYS.ALL_INDEXES.TABLE_OWNER, SYS.ALL_INDEXES.TABLE_NAME,                                                                                                      |
|decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS from                                                                                            |
|SYS.ALL_INDEXES where (SYS.ALL_INDEXES.OWNER in (:1 , :2 , :3 , :4 )                                                                                          |
|and SYS.ALL_INDEXES.TABLE_OWNER in (:5 , :6 , :7 , :8 ) and                                                                                                   |
|SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL' and (SYS.ALL_INDEXES.OWNER,                                                                                             |
|SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((select                                                                                       |
|SYS.ALL_CONSTRAINTS.OWNER, SYS.ALL_CONSTRAINTS.TABLE_NAME,                                                                                                    |
|SYS.ALL_CONSTRAINTS.CONSTRAINT_NAME from SYS.ALL_CONSTRAINTS))) order                                                                                         |
|by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME -- SQL rendered                                                                                          |
|with a free trial version of jOOQ 3.18.0-SNAPSHOT                                                                                                             |
|                                                                                                                                                              |
|Plan hash value: 82967922                                                                                                                                     |
|                                                                                                                                                              |
|---------------------------------------------------------------------------------------------------------------------------------------------------------     |
|| Id  | Operation                                | Name                    | Starts | E-Rows | A-Rows |   A-Time   | Buffers |  OMem |  1Mem | Used-Mem |     |
|---------------------------------------------------------------------------------------------------------------------------------------------------------     |
||   0 | SELECT STATEMENT                         |                         |      1 |        |     24 |00:02:21.88 |    6063K|       |       |          |     |
||   1 |  SORT ORDER BY                           |                         |      1 |      1 |     24 |00:02:21.88 |    6063K|  9216 |  9216 | 8192  (0)|     |
||*  2 |   FILTER                                 |                         |      1 |        |     24 |00:02:21.88 |    6063K|       |       |          |     |
||   3 |    NESTED LOOPS                          |                         |      1 |      1 |   1597 |00:00:00.10 |   14024 |       |       |          |     |
||   4 |     NESTED LOOPS OUTER                   |                         |      1 |      1 |   1597 |00:00:00.08 |   12422 |       |       |          |     |
||   5 |      NESTED LOOPS                        |                         |      1 |      1 |   1597 |00:00:00.08 |   12422 |       |       |          |     |
||   6 |       NESTED LOOPS OUTER                 |                         |      1 |      1 |   2263 |00:00:00.06 |    9656 |       |       |          |     |
||   7 |        NESTED LOOPS OUTER                |                         |      1 |      1 |   2263 |00:00:00.06 |    9648 |       |       |          |     |
||   8 |         NESTED LOOPS OUTER               |                         |      1 |      1 |   2263 |00:00:00.05 |    7803 |       |       |          |     |
||   9 |          NESTED LOOPS                    |                         |      1 |      1 |   2263 |00:00:00.03 |    5536 |       |       |          |     |
||  10 |           NESTED LOOPS                   |                         |      1 |     83 |   2589 |00:00:00.01 |    1363 |       |       |          |     |
||  11 |            INLIST ITERATOR               |                         |      1 |        |      4 |00:00:00.01 |       5 |       |       |          |     |
||  12 |             TABLE ACCESS BY INDEX ROWID  | USER$                   |      4 |      4 |      4 |00:00:00.01 |       5 |       |       |          |     |
||* 13 |              INDEX UNIQUE SCAN           | I_USER1                 |      4 |      4 |      4 |00:00:00.01 |       2 |       |       |          |     |
||* 14 |            INDEX RANGE SCAN              | I_OBJ2                  |      4 |     21 |   2589 |00:00:00.01 |    1358 |       |       |          |     |
||* 15 |           TABLE ACCESS CLUSTER           | IND$                    |   2589 |      1 |   2263 |00:00:00.02 |    4173 |       |       |          |     |
||* 16 |            INDEX UNIQUE SCAN             | I_OBJ#                  |   2589 |      1 |   2589 |00:00:00.01 |    1260 |       |       |          |     |
||  17 |          TABLE ACCESS CLUSTER            | TS$                     |   2263 |      1 |   2263 |00:00:00.01 |    2267 |       |       |          |     |
||* 18 |           INDEX UNIQUE SCAN              | I_TS#                   |   2263 |      1 |   2263 |00:00:00.01 |       4 |       |       |          |     |
||  19 |         TABLE ACCESS CLUSTER             | SEG$                    |   2263 |      1 |    517 |00:00:00.01 |    1845 |       |       |          |     |
||* 20 |          INDEX UNIQUE SCAN               | I_FILE#_BLOCK#          |   2263 |      1 |    517 |00:00:00.01 |    1328 |       |       |          |     |
||* 21 |        INDEX RANGE SCAN                  | I_OBJ1                  |   2263 |      1 |      0 |00:00:00.01 |       8 |       |       |          |     |
||* 22 |       TABLE ACCESS BY INDEX ROWID BATCHED| OBJ$                    |   2263 |      1 |   1597 |00:00:00.02 |    2766 |       |       |          |     |
||* 23 |        INDEX RANGE SCAN                  | I_OBJ1                  |   2263 |      1 |   2263 |00:00:00.01 |    1760 |       |       |          |     |
||* 24 |      INDEX RANGE SCAN                    | I_USER2                 |   1597 |      1 |      0 |00:00:00.01 |       0 |       |       |          |     |
||* 25 |     TABLE ACCESS CLUSTER                 | USER$                   |   1597 |      1 |   1597 |00:00:00.01 |    1602 |       |       |          |     |
||* 26 |      INDEX UNIQUE SCAN                   | I_USER#                 |   1597 |      1 |   1597 |00:00:00.01 |       5 |       |       |          |     |
||* 27 |    FILTER                                |                         |   1597 |        |     77 |00:02:21.69 |    6043K|       |       |          |     |
||  28 |     PARTITION LIST ALL                   |                         |   1597 |     15 |    716 |00:02:21.64 |    6041K|       |       |          |     |
||* 29 |      EXTENDED DATA LINK FULL             | INT$INT$DBA_CONSTRAINTS |   3194 |     15 |    716 |00:02:21.63 |    6041K|       |       |          |     |
||  30 |     NESTED LOOPS SEMI                    |                         |    603 |      1 |      2 |00:00:00.04 |    1840 |       |       |          |     |
||* 31 |      FIXED TABLE FULL                    | X$KZSRO                 |    603 |      2 |   1204 |00:00:00.01 |       0 |       |       |          |     |
||* 32 |      INDEX RANGE SCAN                    | I_OBJAUTH2              |   1204 |      1 |      2 |00:00:00.02 |    1840 |       |       |          |     |
||* 33 |    TABLE ACCESS CLUSTER                  | TAB$                    |   1187 |      1 |   1187 |00:00:00.02 |    2045 |       |       |          |     |
||* 34 |     INDEX UNIQUE SCAN                    | I_OBJ#                  |   1187 |      1 |   1187 |00:00:00.01 |     682 |       |       |          |     |
||  35 |    NESTED LOOPS SEMI                     |                         |   1183 |      1 |     15 |00:00:00.04 |    3652 |       |       |          |     |
||* 36 |     FIXED TABLE FULL                     | X$KZSRO                 |   1183 |      2 |   2351 |00:00:00.02 |       0 |       |       |          |     |
||* 37 |     INDEX RANGE SCAN                     | I_OBJAUTH2              |   2351 |      1 |     15 |00:00:00.01 |    3652 |       |       |          |     |
|---------------------------------------------------------------------------------------------------------------------------------------------------------     |
|                                                                                                                                                              |
|Predicate Information (identified by operation id):                                                                                                           |
|---------------------------------------------------                                                                                                           |
|                                                                                                                                                              |
|   2 - filter(( IS NULL AND ("IO"."TYPE#"<>2 OR =1) AND ("IO"."OWNER#"=USERENV('SCHEMAID') OR                                                                 |
|              ORA_CHECK_SYS_PRIVILEGE("IO"."OWNER#","IO"."TYPE#")=1 OR  IS NOT NULL)))                                                                        |
|  13 - access(("IU"."NAME"=:5 OR "IU"."NAME"=:6 OR "IU"."NAME"=:7 OR "IU"."NAME"=:8))                                                                         |
|  14 - access("IO"."OWNER#"="IU"."USER#" AND "IO"."TYPE#"=2)                                                                                                  |
|       filter("IO"."TYPE#"=2)                                                                                                                                 |
|  15 - filter((INTERNAL_FUNCTION("I"."TYPE#") AND DECODE(BITAND("I"."PROPERTY",16),0,'','FUNCTION-BASED                                                       |
|              ')||DECODE("I"."TYPE#",1,'NORMAL'||DECODE(BITAND("I"."PROPERTY",4),0,'',4,'/REV'),2,'BITMAP',3,'CLUSTER',4,'IOT - TOP',5,'IOT -                 |
|              NESTED',6,'SECONDARY',7,'ANSI',8,'LOB',9,'DOMAIN')='NORMAL' AND BITAND("I"."FLAGS",4096)=0))                                                    |
|  16 - access("I"."BO#"="IO"."OBJ#")                                                                                                                          |
|  18 - access("I"."TS#"="TS"."TS#")                                                                                                                           |
|  20 - access("I"."TS#"="S"."TS#" AND "I"."FILE#"="S"."FILE#" AND "I"."BLOCK#"="S"."BLOCK#")                                                                  |
|  21 - access("I"."INDMETHOD#"="ITO"."OBJ#")                                                                                                                  |
|  22 - filter(BITAND("O"."FLAGS",128)=0)                                                                                                                      |
|  23 - access("O"."OBJ#"="I"."OBJ#")                                                                                                                          |
|  24 - access("ITO"."OWNER#"="ITU"."USER#")                                                                                                                   |
|  25 - filter(("U"."NAME"=:1 OR "U"."NAME"=:2 OR "U"."NAME"=:3 OR "U"."NAME"=:4))                                                                             |
|  26 - access("U"."USER#"="O"."OWNER#")                                                                                                                       |
|  27 - filter(("OWNER"=SYS_CONTEXT('USERENV','CURRENT_USER') OR ORA_CHECK_SYS_PRIVILEGE("OWNERID","OBJECT_TYPE#")=1 OR  IS NOT NULL))                         |
|  29 - filter((("INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=4 OR ("INT$INT$DBA_CONSTRAINTS"."OBJECT_TYPE#"=2 AND                                                 |
|              "INT$INT$DBA_CONSTRAINTS"."ORIGIN_CON_ID"=TO_NUMBER(SYS_CONTEXT('USERENV','CON_ID')))) AND LNNVL("OWNER"<>:B1) AND LNNVL("TABLE_NAME"<>:B2) AND |
|              LNNVL("CONSTRAINT_NAME"<>:B3)))                                                                                                                 |
|  31 - filter(("CON_ID"=0 OR "CON_ID"=3))                                                                                                                     |
|  32 - access("GRANTEE#"="KZSROROL" AND "OBJ#"=OBJ_ID(:B1,:B2,:B3,:B4))                                                                                       |
|  33 - filter(BITAND("T"."PROPERTY",36893488147419103232)=0)                                                                                                  |
|  34 - access("T"."OBJ#"=:B1)                                                                                                                                 |
|  36 - filter(("CON_ID"=0 OR "CON_ID"=3))                                                                                                                     |
|  37 - access("GRANTEE#"="KZSROROL" AND "OBJ#"=:B1)                                                                                                           |
|                                                                                                                                                              |
```

**Comment by lukaseder** [comment]
Manually turning `NOT IN` into `NOT EXISTS` yields no results

**Comment by lukaseder** [comment]
Adding an additional predicate filtering on `ALL_CONSTRAINTS.OWNER` also doesn't help

**Comment by lukaseder** [comment]
This could be a solution:

```sql
with cas (
  select /*+MATERIALIZE*/ SYS.ALL_CONSTRAINTS.OWNER, SYS.ALL_CONSTRAINTS.TABLE_NAME, SYS.ALL_CONSTRAINTS.CONSTRAINT_NAME
  from SYS.ALL_CONSTRAINTS
  where SYS.ALL_CONSTRAINTS.OWNER in (
    'TEST', 'MULTI_SCHEMA', 'USR_2522', 'SYS'
  )
)
select
  SYS.ALL_INDEXES.OWNER,
  SYS.ALL_INDEXES.INDEX_NAME,
  SYS.ALL_INDEXES.TABLE_OWNER,
  SYS.ALL_INDEXES.TABLE_NAME,
  decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS
from SYS.ALL_INDEXES
where (
  SYS.ALL_INDEXES.OWNER in (
    'TEST', 'MULTI_SCHEMA', 'USR_2522', 'SYS'
  )
  and SYS.ALL_INDEXES.TABLE_OWNER in (
    'TEST', 'MULTI_SCHEMA', 'USR_2522', 'SYS'
  )
  and SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL'
  and (SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((
    select owner, table_name, constraint_name
    from c
  ))
)
order by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.5 (https://github.com/jOOQ/jOOQ/issues/14015), and 3.16.11 (https://github.com/jOOQ/jOOQ/issues/14016)
